File: src/demos/irrlicht/demo_IRR_rot_spring.cpp
// =============================================================================
// PROJECT CHRONO - http://projectchrono.org
//
// Copyright (c) 2014 projectchrono.org
// All rights reserved.
//
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file at the top level of the distribution and at
// http://projectchrono.org/license-chrono.txt.
//
// =============================================================================
// Authors: Radu Serban
// =============================================================================
//
// Simple example demonstrating the use of ChLinkRotSpringCB.
//
// Recall that Irrlicht uses a left-hand frame, so everything is rendered with
// left and right flipped.
//
// =============================================================================

#include <cstdio>

#include "chrono/physics/ChSystemNSC.h"
#include "chrono/physics/ChBody.h"

#include "chrono_irrlicht/ChIrrApp.h"

using namespace chrono;
using namespace chrono::irrlicht;
using namespace irr;

// =============================================================================

double spring_coef = 40;
double damping_coef = 2;
double rest_angle = CH_C_PI / 6;

// =============================================================================

// Functor class implementing the torque for a ChLinkRotSpringCB link.
class MySpringTorque : public ChLinkRotSpringCB::TorqueFunctor {
    virtual double operator()(double time,             // current time
                              double angle,            // relative angle of rotation
                              double vel,              // relative angular speed
                              ChLinkRotSpringCB* link  // back-pointer to associated link
                              ) override {
        double torque = -spring_coef * (angle - rest_angle) - damping_coef * vel;
        return torque;
    }
};

// =============================================================================

int main(int argc, char* argv[]) {
    GetLog() << "Copyright (c) 2017 projectchrono.org\nChrono version: " << CHRONO_VERSION << "\n\n";

    ChSystemNSC system;
    system.Set_G_acc(ChVector<>(0, 0, 0));

    //ChQuaternion<> rev_rot = QUNIT;
    ChQuaternion<> rev_rot = Q_from_AngX(CH_C_PI / 6.0);
    //ChQuaternion<> rev_rot = Q_from_AngX(CH_C_PI / 2.0);

    ChVector<> rev_dir = rev_rot.GetZaxis();

    ChVector<> rev_pos(+1, 0, 0);

    // Create ground body
    auto ground = chrono_types::make_shared<ChBody>();
    system.AddBody(ground);
    ground->SetIdentifier(-1);
    ground->SetBodyFixed(true);
    ground->SetCollide(false);

    // Visualization for revolute joint
    auto cyl_rev = chrono_types::make_shared<ChCylinderShape>();
    cyl_rev->GetCylinderGeometry().p1 = rev_pos + 0.2 * rev_dir;
    cyl_rev->GetCylinderGeometry().p2 = rev_pos - 0.2 * rev_dir;
    cyl_rev->GetCylinderGeometry().rad = 0.1;
    ground->AddAsset(cyl_rev);

    // Offset from joint to body COM
    ChVector<> offset(1.5, 0, 0);

    // Consistent initial velocities
    double omega = 5.0;
    ChVector<> ang_vel = omega * rev_dir;
    ChVector<> lin_vel = Vcross(ang_vel, offset);

    // Create pendulum body
    auto body = chrono_types::make_shared<ChBody>();
    system.AddBody(body);
    body->SetPos(rev_pos + offset);
    body->SetPos_dt(lin_vel);
    body->SetWvel_par(ang_vel);
    body->SetIdentifier(1);
    body->SetBodyFixed(false);
    body->SetCollide(false);
    body->SetMass(1);
    body->SetInertiaXX(ChVector<>(1, 1, 1));

    // Attach visualization assets
    auto sph = chrono_types::make_shared<ChSphereShape>();
    sph->GetSphereGeometry().rad = 0.3;
    body->AddAsset(sph);
    auto cyl = chrono_types::make_shared<ChCylinderShape>();
    cyl->GetCylinderGeometry().p1 = ChVector<>(-1.5, 0, 0);
    cyl->GetCylinderGeometry().p2 = ChVector<>(0, 0, 0);
    cyl->GetCylinderGeometry().rad = 0.1;
    body->AddAsset(cyl);
    auto col = chrono_types::make_shared<ChColorAsset>();
    col->SetColor(ChColor(0.7f, 0.8f, 0.8f));
    body->AddAsset(col);

    // Create revolute joint between body and ground
    auto rev = chrono_types::make_shared<ChLinkLockRevolute>();
    rev->Initialize(body, ground, ChCoordsys<>(rev_pos, rev_rot));
    system.AddLink(rev);

    // Create the rotational spring between body and ground
    auto torque = chrono_types::make_shared<MySpringTorque>();
    auto spring = chrono_types::make_shared<ChLinkRotSpringCB>();
    spring->Initialize(body, ground, ChCoordsys<>(rev_pos, rev_rot));
    spring->RegisterTorqueFunctor(torque);
    system.AddLink(spring);

    // Create the Irrlicht application
    ChIrrApp application(&system, L"ChLinkRotSpringCB demo", core::dimension2d<u32>(800, 600), false, true);
    application.AddTypicalLogo();
    application.AddTypicalSky();
    application.AddTypicalLights();
    application.AddTypicalCamera(core::vector3df(3, 1, 3));

    application.AssetBindAll();
    application.AssetUpdateAll();

    // Simulation loop
    int frame = 0;

    application.SetTimestep(0.001);

    GetLog().SetNumFormat("%10.3f");

    while (application.GetDevice()->run()) {
        application.BeginScene();
        application.DrawAll();
        ChIrrTools::drawAllCOGs(system, application.GetVideoDriver(), 1.0);
        ChIrrTools::drawAllLinkframes(system, application.GetVideoDriver(), 1.5);
        application.EndScene();
        application.DoStep();

        if (frame % 50 == 0) {
            GetLog() << system.GetChTime() << "\n";
            GetLog() << "Body position" << body->GetPos() << "\n";
            GetLog() << "Body lin. vel." << body->GetPos_dt() << "\n";
            GetLog() << "Body absolute ang. vel." << body->GetWvel_par() << "\n";
            GetLog() << "Body local ang. vel." << body->GetWvel_loc() << "\n";
            GetLog() << "Rot. spring-damper  " << spring->GetRotSpringAngle() << "  " << spring->GetRotSpringSpeed()
                     << "  " << spring->GetRotSpringTorque() << "\n";
            GetLog() << "---------------\n\n";
        }

        frame++;
    }

    return 0;
}

File: src/demos/python/irrlicht/demo_IRR_rot_spring.py
# =============================================================================
# PROJECT CHRONO - http://projectchrono.org
#
# Copyright (c) 2014 projectchrono.org
# All rights reserved.
#
# Use of this source code is governed by a BSD-style license that can be found
# in the LICENSE file at the top level of the distribution and at
# http://projectchrono.org/license-chrono.txt.
#
# =============================================================================
# Authors: Radu Serban
# =============================================================================
#
# Simple example demonstrating the use of ChLinkRSDA.
#
# Recall that Irrlicht uses a left-hand frame, so everything is rendered with
# left and right flipped.
#
# =============================================================================

import pychrono as chrono
import pychrono.irrlicht as irr
import math as m

# =============================================================================

spring_coef = 40
damping_coef = 2
rest_angle = m.pi / 6

# =============================================================================

# Functor class implementing the torque for a ChLinkRSDA link.
class MySpringTorque(chrono.TorqueFunctor):
    def __init__(self):
        super(MySpringTorque, self).__init__()
    
    def evaluate(self,    #
                 time,    # current time
                 angle,   # relative angle of rotation
                 vel,     # relative angular speed
                 link):   # associated link
                              
        torque = -spring_coef * (angle - rest_angle) - damping_coef * vel
        return torque

# =============================================================================

print("Copyright (c) 2017 projectchrono.org")

sys = chrono.ChSystemNSC()
sys.Set_G_acc(chrono.ChVectorD(0, 0, 0))

# Revolute joint frame 
rev_rot = chrono.Q_from_AngX(m.pi / 6.0)
rev_dir = rev_rot.GetZaxis()
rev_pos = chrono.ChVectorD(+1, 0, 0)

# Create ground body
ground = chrono.ChBody()
sys.AddBody(ground)
ground.SetIdentifier(-1)
ground.SetBodyFixed(True)
ground.SetCollide(False)

# Visualization for revolute joint
cyl_rev = chrono.ChCylinderShape()
cyl_rev.GetCylinderGeometry().p1 = rev_pos + rev_dir * 0.2
cyl_rev.GetCylinderGeometry().p2 = rev_pos - rev_dir * 0.2
cyl_rev.GetCylinderGeometry().rad = 0.1
ground.AddVisualShape(cyl_rev)

# Offset from joint to body COM
offset = chrono.ChVectorD(1.5, 0, 0)

# Consistent initial velocities
omega = 5.0
ang_vel = rev_dir * omega
lin_vel = ang_vel % offset

# Create pendulum body
body = chrono.ChBody()
sys.AddBody(body)
body.SetPos(rev_pos + offset)
body.SetPos_dt(lin_vel)
body.SetWvel_par(ang_vel)
body.SetIdentifier(1)
body.SetBodyFixed(False)
body.SetCollide(False)
body.SetMass(1)
body.SetInertiaXX(chrono.ChVectorD(1, 1, 1))

# Attach visualization assets
sph = chrono.ChSphereShape()
sph.GetSphereGeometry().rad = 0.3
body.AddVisualShape(sph)
cyl = chrono.ChCylinderShape()
cyl.GetCylinderGeometry().p1 = chrono.ChVectorD(-1.5, 0, 0)
cyl.GetCylinderGeometry().p2 = chrono.ChVectorD(0, 0, 0)
cyl.GetCylinderGeometry().rad = 0.1
cyl.SetColor(chrono.ChColor(0.7, 0.8, 0.8))
body.AddVisualShape(cyl)

# Create revolute joint between body and ground
rev = chrono.ChLinkLockRevolute()
rev.Initialize(body, ground, chrono.ChCoordsysD(rev_pos, rev_rot))
sys.AddLink(rev)

# Create the rotational spring between body and ground
torque = MySpringTorque()
spring = chrono.ChLinkRSDA()
spring.Initialize(body, ground, chrono.ChCoordsysD(rev_pos, rev_rot))
spring.RegisterTorqueFunctor(torque)
rsda = chrono.ChRotSpringShape(0.5, 40)
rsda.SetColor(chrono.ChColor(0, 0, 0))
spring.AddVisualShape(rsda)
sys.AddLink(spring)

# Create the Irrlicht visualization
vis = irr.ChVisualSystemIrrlicht()
sys.SetVisualSystem(vis)
vis.SetWindowSize(1024,768)
vis.SetWindowTitle('ChLinkRSDA demo')
vis.Initialize()
vis.AddLogo(chrono.GetChronoDataFile('logo_pychrono_alpha.png'))
vis.AddSkyBox()
vis.AddCamera(chrono.ChVectorD(3, 1, 3))
vis.AddTypicalLights()

# Simulation loop
frame = 0
while vis.Run():
    vis.BeginScene() 
    vis.DrawAll()
    irr.drawAllCOGs(sys, vis.GetVideoDriver(), 1.0)
    irr.drawAllLinkframes(sys, vis.GetVideoDriver(), 1.5)
    vis.EndScene()
    sys.DoStepDynamics(1e-3)

    if (frame % 50 == 0) :
        print('{:.6}'.format(str(sys.GetChTime())))
        print('Body position      ', body.GetPos())
        print('Body lin. vel      ', body.GetPos_dt())
        print('Body abs. ang. vel ', body.GetWvel_par())
        print('Body loc. ang. vel ', body.GetWvel_loc())
        print('Rot. spring-damper ', spring.GetAngle(), '  ', spring.GetTorque())
        print('---------------')


    frame += 1
